lib/Backend/Lower.cpp

```diff
@@ -19489,6 +19535,57 @@ void Lowerer::GenerateFastInlineIsIn(IR::Instr * instr)
     instr->InsertAfter(doneLabel);
 }
 
+// Generate a fast path for the "in" operator to use the cache where the key may be a PropertyString or Symbol.
+void Lowerer::GenerateFastObjectIsIn(IR::Instr * instr)
+{
+    if (!instr->GetSrc1()->IsRegOpnd() || !instr->GetSrc2()->IsRegOpnd())
+    {
+        // REVIEW: Is it worth adding logic for this case? Under what circumstances would the operands not be registers?
+        return;
+    }
+
+    IR::RegOpnd* baseOpnd = instr->GetSrc2()->AsRegOpnd();
+
+    if (!baseOpnd->GetValueType().IsLikelyObject())
+    {
+        return;
+    }
+
+    IR::RegOpnd* indexOpnd = instr->GetSrc1()->AsRegOpnd();
+    IR::LabelInstr* helperLabel = IR::LabelInstr::New(Js::OpCode::Label, m_func, true);
+    IR::LabelInstr* doneLabel = IR::LabelInstr::New(Js::OpCode::Label, m_func);
+    bool addLabels = true;
+
+    if (indexOpnd->GetValueType().IsLikelyString())
+    {
+        GeneratePropertyStringTest(indexOpnd, instr, helperLabel, true /*usePoison*/);
+
+        const uint32 inlineCacheOffset = Js::PropertyString::GetOffsetOfLdElemInlineCache();
+        const uint32 hitRateOffset = Js::PropertyString::GetOffsetOfHitRate();
+
+        GenerateFastIsInSymbolOrStringIndex(instr, indexOpnd, baseOpnd, instr->GetDst(), inlineCacheOffset, hitRateOffset, helperLabel, doneLabel);
+    }
+    else if (indexOpnd->GetValueType().IsLikelySymbol())
+    {
+        GenerateSymbolTest(indexOpnd, instr, helperLabel);
+
+        const uint32 inlineCacheOffset = Js::JavascriptSymbol::GetOffsetOfLdElemInlineCache();
+        const uint32 hitRateOffset = Js::JavascriptSymbol::GetOffsetOfHitRate();
+
+        GenerateFastIsInSymbolOrStringIndex(instr, indexOpnd, baseOpnd, instr->GetDst(), inlineCacheOffset, hitRateOffset, helperLabel, doneLabel);
+    }
+    else
+    {
+        addLabels = false;
+    }
+
+    if (addLabels)
+    {
```

Comment:
nit: We can early return if its not likely symbol/string ?

Reply:
Good point, thanks.

---
In reply to: [175843556](https://github.com/Microsoft/ChakraCore/pull/4846#discussion_r175843556) [](ancestors = 175843556)